core/src/test/scala/com.mediative.sparrow/DataFrameReaderTest.scala

```diff
@@ -173,51 +195,107 @@ class DataFrameReaderTest extends FreeSpec with BeforeAndAfterAll {
       assert(toRDD[T](df) == expected.failure)
     }
 
-    // To get DataFrame#toRDD usage.
-    import com.mediative.sparrow.syntax.df._
-
-    "round-trip an object containing an Int and a String from RDD to DataFrame back to RDD" in {
-      import TestCaseClasses.TestToRdd1
-
-      val expected = TestToRdd1(1, "a")
-      val df = sqlContext.createDataFrame(sc.parallelize(List(expected)))
+    "successfully marshall RDD => DataFrame => RDD an object containing" - {
+      // To get DataFrame#toRDD usage.
+      import com.mediative.sparrow.syntax.df._
+      import scala.reflect.runtime.universe.TypeTag
+      def testRDD_DataFrame_codec[T <: Product: ClassTag: TypeTag: RowConverter](value: T): Unit = {
+        val rdd0 = sc.parallelize(List(value))
+        assertResult(1) { rdd0.count }
+        val df = sqlContext.createDataFrame(rdd0)
+        val rdd1Maybe = df.toRDD[T]
+        assert(rdd1Maybe.isSuccess, rdd1Maybe)
+        val rdd1 = rdd1Maybe.toOption.get
+        assertResult(0) { rdd0.subtract(rdd1).count }
+        assertResult(0) { rdd1.subtract(rdd0).count }
+      }
 
-      assert(df.toRDD[TestToRdd1].toOption.get.first == expected)
-    }
+      import TestCaseClasses._
 
-    "round-trip an object containing an Int and an optional Int from RDD to DataFrame back to RDD" - {
-      "when containing some value" in {
-        import TestCaseClasses.TestToRdd2
+      "Int, String" in {
+        testRDD_DataFrame_codec(TestToRdd1(1, "a"))
+      }
 
-        val expected = TestToRdd2(1, Option(1))
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "Int, Option[Int]" - {
+        "when Some(Int)" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option(1)))
+        }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+        "when None" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option.empty))
+        }
       }
 
-      "when containing no value" in {
-        import TestCaseClasses.TestToRdd2
-
-        val expected = TestToRdd2(1, Option.empty)
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "String, java.sql.Timestamp" in {
+        pendingUntilFixed {
+          // FIXME:
+          // rdd1Maybe.isSuccess was false Failure(NonEmptyList(The field 'timestampVal' isn't a LongType as expected, TimestampType received.)) (DataFrameReaderTest.scala:203)
+          testRDD_DataFrame_codec(
+            TestToRdd3("a", Timestamp.valueOf("2015-07-15 09:00:00"))
+          )
+        }
+      }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+      "Int, Double" in {
+        pendingUntilFixed {
+          // FIXME:
+          // "org.apache.spark.SparkException: Job aborted due to stage failure"
+          // Caused by: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
```

Comment:
The bug fix in Parquet I've no idea how realistic it would be to have fixed. But have you tried if it even fails in Scala 2.11 with the latest version of Spark and Hadoop? (using `sbt + test`)

But anyway, the real fix I think is having Sparrow support conversion in both direction, `Row` => `T` and `T` => `Row`. It's what I'm trying to do right now, figure out a design for it.